xarray/core/duck_array_ops.py

```diff
@@ -278,6 +276,22 @@ def f(values, axis=None, skipna=None, **kwargs):
 cumsum_1d.numeric_only = True
 
 
+_mean = _create_nan_agg_method('mean')
+
+def mean(array, axis=None, skipna=None, **kwargs):
+    """ inhouse mean that can handle datatime dtype """
+    array = asarray(array)
+    if array.dtype.kind == 'M':
+        offset = min(array)
+        dtype = (array.ravel()[0] - offset).dtype  # dtype.kind == 'm'
```

Comment:
This will fail for size 0 arrays. Maybe switch to another reduce method if the array is size 0, e.g., use min instead? The only important part is that it handles reducing the shape properly.

Reply:
Fixed, but I am not yet sure it (the updated one) is the best solution to find the appropriate dtype. See [comment](https://github.com/pydata/xarray/pull/2612/files/82c08afdaf38ab0785daca73f5c882a1ce320227#diff-b81ae485ac6616ab60b4c70334009571)